### PR TITLE
enable const-float-classify test, and test_next_up/down on 32bit x86

### PR DESCRIPTION
**Pull request by RalfJung**
The  test_next_up/down tests have been disabled on all 32bit x86 targets, which goes too far -- they should definitely work on our (tier 1) i686 target, it is only without SSE that we might run into trouble due to https://github.com/rust-lang/rust/issues/114479. However, I cannot reproduce that trouble any more -- maybe that got fixed by https://github.com/rust-lang/rust/pull/123351?

The  const-float-classify test relied on const traits "because we can", and got disabled when const traits got removed. That's an unfortunate reduction in test coverage of our float functionality, so let's restore the test in a way that does not rely on const traits.

The const-float tests are actually testing runtime behavior as well, and I don't think that runtime behavior is covered anywhere else. Probably they shouldn't be called "const-float", but we don't have a `tests/ui/float` folder... should I create one and move them there? Are there any other ui tests that should be moved there?

I also removed some FIXME referring to not use x87 for Rust-to-Rust-calls -- that has happened in #123351 so this got fixed indeed. Does that mean we can simplify all that float code again? I am not sure how to test it. Is running the test suite with an i586 target enough?

Cc @tgross35 @workingjubilee

### CODE REVIEW

**Comment by RalfJung** [comment]
I tried this diff:
```diff
diff --git a/library/core/src/num/f32.rs b/library/core/src/num/f32.rs
index 4c089946033..ca2205d38e1 100644
--- a/library/core/src/num/f32.rs
+++ b/library/core/src/num/f32.rs
@@ -662,13 +662,7 @@ pub const fn classify(self) -> FpCategory {
         // hardware flushes subnormals to zero. These are platforms bugs, and Rust will misbehave on
         // such hardware, but we can at least try to make things seem as sane as possible by being
         // careful here.
-        if self.is_infinite() {
-            // A value may compare unequal to infinity, despite having a "full" exponent mask.
-            FpCategory::Infinite
-        } else if self.is_nan() {
-            // And it may not be NaN, as it can simply be an "overextended" finite value.
-            FpCategory::Nan
-        } else {
+        {
             // However, std can't simply compare to zero to check for zero, either,
             // as correctness requires avoiding equality tests that may be Subnormal == -0.0
             // because it may be wrong under "denormals are zero" and "flush to zero" modes.
@@ -683,6 +677,8 @@ pub const fn classify(self) -> FpCategory {
             // I fear losing mantissa data that would have answered that differently.
             let b = self.to_bits();
             match (b & Self::MAN_MASK, b & Self::EXP_MASK) {
+                (0, Self::EXP_MASK) => FpCategory::Infinite,
+                (_, Self::EXP_MASK) => FpCategory::Nan,
                 (0, 0) => FpCategory::Zero,
                 (_, 0) => FpCategory::Subnormal,
                 _ => FpCategory::Normal,
diff --git a/library/core/src/num/f64.rs b/library/core/src/num/f64.rs
index bcef8fe7724..b5ef9e85aac 100644
--- a/library/core/src/num/f64.rs
+++ b/library/core/src/num/f64.rs
@@ -663,9 +663,7 @@ pub const fn classify(self) -> FpCategory {
         //
         // Thus, a value may compare unequal to infinity, despite having a "full" exponent mask.
         // And it may not be NaN, as it can simply be an "overextended" finite value.
-        if self.is_nan() {
-            FpCategory::Nan
-        } else {
+        {
             // However, std can't simply compare to zero to check for zero, either,
             // as correctness requires avoiding equality tests that may be Subnormal == -0.0
             // because it may be wrong under "denormals are zero" and "flush to zero" modes.
@@ -678,6 +676,7 @@ pub const fn classify(self) -> FpCategory {
             let b = self.to_bits();
             match (b & Self::MAN_MASK, b & Self::EXP_MASK) {
                 (0, Self::EXP_MASK) => FpCategory::Infinite,
+                (_, Self::EXP_MASK) => FpCategory::Nan,
                 (0, 0) => FpCategory::Zero,
                 (_, 0) => FpCategory::Subnormal,
                 _ => FpCategory::Normal,
```
And the tests seem to still pass. But I am not sure if there are more tests you ran, @workingjubilee ? Is it worth trying the bit-based classify implementation again?

I think I'll rather not touch this code for now.^^

**Comment by tgross35** [comment]
> The test_next_up/down tests have been disabled on all 32bit x86 targets, which goes too far -- they should definitely work on our (tier 1) i686 target, it is only without SSE that we might run into trouble due to #114479. However, I cannot reproduce that trouble any more -- maybe that got fixed by #123351?

Could a similar test be tried making sure the calls are `extern "C"`? I suspect that https://github.com/rust-lang/rust/pull/123351 did fix things for the default ABI, but it left C untouched.

**Comment by tgross35** [comment]
> The const-float tests are actually testing runtime behavior as well, and I don't think that runtime behavior is covered anywhere else. Probably they shouldn't be called "const-float", but we don't have a `tests/ui/float` folder... should I create one and move them there? Are there any other ui tests that should be moved there?

I don't really know what a good solution is here. We have a ton of float tests that are quite similar but _slightly_ different, spread out in a bunch of different locations - `std` tests, UI tests, miri, compiler_builtins, etc. Even the pattern generation we have at https://github.com/rust-lang/rust/tree/a7399ba69d37b019677a9c47fe89ceb8dd82db2d/src/etc/test-float-parse now would be pretty useful for making sure the edge cases are checked.

I know that is entirely unhelpful, just that it would be nice if we had a single suite we could run against both const and runtime somehow.

**Comment by RalfJung** [comment]
> Could a similar test be tried making sure the calls are extern "C"? I suspect that https://github.com/rust-lang/rust/pull/123351 did fix things for the default ABI, but it left C untouched.

This calls a bunch of Rust functions though so the Rust ABI is what we want. What would be the point of using the C ABI here? We know that one is broken...

**Comment by tgross35** [comment]
Ah, I thought you were still trying to work around the fix to repro the behavior that led to these notes.

**Comment by RalfJung** [comment]
I tried adding extern "C" to from_bits, next_up, next_down -- that made no difference.

So yeah not sure how to reproduce the failures. I guess we should do a try-build of the relevant runner. Time to figure out how that works...

**Comment by RalfJung** [comment]
These tests seem to have been disabled on x86 since they got added: https://github.com/rust-lang/rust/pull/100578. And that [goes back](https://github.com/rust-lang/rust/pull/88728#issuecomment-950049318) to [this CI failure](https://github.com/rust-lang/rust/pull/90178#issuecomment-950026477). So we should be seeing it on CI if it's still a problem.

**Comment by RalfJung** [comment]
But the tests passed on CI. So... maybe something changed in LLVM that makes this less of a problem in practice?

**Comment by RalfJung** [comment]
The job passed, and then the "dist" part failed of course.

So... something else must have changed to make the text_up/down stuff work on x87 now.

@workingjubilee how did you test the `classify` stuff? Do you have a set that can reproduce a failure when one does "naive" bit-based classification?

**Comment by RalfJung** [comment]
Anyway that seems orthogonal to this PR -- test_next_up/down were disabled on x86 because they failed on CI; they no longer fail on CI; that should be sufficient to enable these tests again.

File: tests/ui/float/float-classify-runtime-const.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      1.0 / 0.0 => [ Infinite,  false,        true,     false,     false,             true,            false]
+     -1.0 / 0.0 => [ Infinite,  false,        true,     false,     false,            false,             true]
+   1.0 / T::MAX => [Subnormal,  false,       false,      true,     false,             true,            false]
+  -1.0 / T::MAX => [Subnormal,  false,       false,      true,     false,            false,             true]
+}
+
+fn main() {
+    f32();
+    f64();
+    // FIXME(f16_f128): also test f16 and f128
```
Comment (tgross35) [inline tests/ui/float/float-classify-runtime-const.rs:80]: Ty :)

File: tests/ui/float/float-classify-runtime-const.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      0.0 / 0.0 => [      Nan,   true,       false,     false,     false,           NonDet,           NonDet]
+            1.0 => [   Normal,  false,       false,      true,      true,             true,            false]
+           -1.0 => [   Normal,  false,       false,      true,      true,            false,             true]
+            0.0 => [     Zero,  false,       false,      true,     false,             true,            false]
+           -0.0 => [     Zero,  false,       false,      true,     false,            false,             true]
+      1.0 / 0.0 => [ Infinite,  false,        true,     false,     false,             true,            false]
+     -1.0 / 0.0 => [ Infinite,  false,        true,     false,     false,            false,             true]
+   1.0 / T::MAX => [Subnormal,  false,       false,      true,     false,             true,            false]
+  -1.0 / T::MAX => [Subnormal,  false,       false,      true,     false,            false,             true]
+}
```
Comment (RalfJung) [inline tests/ui/float/float-classify-runtime-const.rs:75]: What changed in this test is the addition of the `classify` column, and the two rows at the end that test a subnormal.

**Comment by Nadrieril** [comment]
not an area I know anything about

r? @tgross35

**Comment by Nadrieril** [comment]
r? compiler

**Comment by tgross35** [comment]
Probably 

r? @workingjubilee 

To confirm why some of this was failing in the first place. Cc @orlp too since it looks like you did the initial implementation here, in case you remember anything (quite possibly not, considering it was two years ago :)

**Comment by orlp** [comment]
@tgross35 At least when I did the implementation, i686 normalized floats (deleting NaN payloads, flushing subnormals to zero, perhaps other stuff) at function boundaries, which messed with the test suite.

For [example](https://github.com/rust-lang/rust/issues/91399#issuecomment-1045989643), I found that `x.to_bits() == x.to_bits()` can succeed while `identity(x).to_bits() == x.to_bits()` can fail when `identity` isn't inlined, even if `identity` is defined as `fn identity(x: f64) -> f64 { x }`.

**Comment by workingjubilee** [comment]
I believe LLVM has changed how it compiles floats in ways that significantly affect our outputs here, so it's not at all surprising to me that we're having trouble replicating the original regression we were protecting against.

**Comment by RalfJung** [comment]
Flushing denormals also sounds odd, AFAIK that is not part of the usual behavior of moving things to an x87 register? According to @beetrees  [here](https://github.com/rust-lang/rust/issues/115567#issuecomment-2182588861), only the quiet bit should change.

**Comment by orlp** [comment]
I may be misremembering things, that's also possible. What was a fact was that those disabled platforms were creating failures for what seemed to be reasonable code that passed the test suite on the modern platforms, so the advice I got was to just skip the tests on those platforms.

**Comment by RalfJung** [comment]
I think the relevant CI failure from back then is [this one](https://github.com/rust-lang/rust/pull/90178#issuecomment-950026477). So it's about the i586 target, i.e. the one without SSE.

**Comment by RalfJung** [comment]
> when compiled with optimisations for i586-unknown-linux-gnu, will emit the nonsensical:

Oh, "fun".
Though that seems like `classify` is actually doing it right and `is_infinite` should be adjusted. And `classify` does use the bit value to check for infinity on `f64`.

So... my take-away from that is that when/if we touch the existing work-arounds in `classify` we should add a bunch more cases to the float-bits-conv test, if we care about things working on i586. (And we have a non-trivial amount of code dedicated to that, also e.g. in dec2flt handling where we even update the FPU precision to ensure consistent results on i586.) But I don't think that should affect this PR.

File: library/core/src/num/f16.rs [hunk 3]
```diff
@@ -435,6 +435,7 @@ impl f16 {
         // WASM, see llvm/llvm-project#96437). These are platforms bugs, and Rust will misbehave on
         // such platforms, but we can at least try to make things seem as sane as possible by being
         // careful here.
+        // Cc https://github.com/rust-lang/rust/issues/114479
```
Comment (workingjubilee) [inline library/core/src/num/f16.rs:438]: ...carbon copy?

shouldn't this be ["quod vide" (qv.)](https://en.wikipedia.org/wiki/List_of_Latin_phrases_(full)#quod_vide) or ["conferatur" (cf.)](https://en.wikipedia.org/wiki/Cf.)? or just "see also"?
```suggestion
        // see also https://github.com/rust-lang/rust/issues/114479
```

File: library/core/src/num/f32.rs [hunk 4]
```diff
@@ -662,10 +662,7 @@ impl f32 {
         // hardware flushes subnormals to zero. These are platforms bugs, and Rust will misbehave on
         // such hardware, but we can at least try to make things seem as sane as possible by being
         // careful here.
-        //
-        // FIXME(jubilee): Using x87 operations is never necessary in order to function
-        // on x86 processors for Rust-to-Rust calls, so this issue should not happen.
-        // Code generation should be adjusted to use non-C calling conventions, avoiding this.
+        // Cc https://github.com/rust-lang/rust/issues/114479
```
Comment (workingjubilee) [inline library/core/src/num/f32.rs:665]:
```suggestion
        // see also https://github.com/rust-lang/rust/issues/114479
```

File: library/core/src/num/f64.rs [hunk 5]
```diff
@@ -660,10 +660,7 @@ impl f64 {
         // float semantics Rust relies on: x87 uses a too-large exponent, and some hardware flushes
         // subnormals to zero. These are platforms bugs, and Rust will misbehave on such hardware,
         // but we can at least try to make things seem as sane as possible by being careful here.
-        //
-        // FIXME(jubilee): Using x87 operations is never necessary in order to function
-        // on x86 processors for Rust-to-Rust calls, so this issue should not happen.
-        // Code generation should be adjusted to use non-C calling conventions, avoiding this.
+        // Cc https://github.com/rust-lang/rust/issues/114479
```
Comment (workingjubilee) [inline library/core/src/num/f64.rs:663]:
```suggestion
        // see also https://github.com/rust-lang/rust/issues/114479
```

File: tests/ui/numbers-arithmetic/issue-105626.rs [hunk 6]
```diff
@@ -11,6 +11,7 @@ fn main() {
     assert_ne!((n as f64) as f32, n as f32);
 
     // FIXME: these assertions fail if only x87 is enabled
+    // Cc https://github.com/rust-lang/rust/issues/114479
```
Comment (workingjubilee) [inline tests/ui/numbers-arithmetic/issue-105626.rs:14]:
```suggestion
    // see also https://github.com/rust-lang/rust/issues/114479
```

**Review by workingjubilee — APPROVED** [review]
r=me with either the intended Latin (abbreviated or no) or an English phrase that would take a document instead of a person as its object. :^)

**Comment by RalfJung** [comment]
Good point. :)

@bors r=workingjubilee

**Comment by workingjubilee** [comment]
@RalfJung I have rebased your PR, I hope that's okay.

**Comment by RalfJung** [comment]
> As a minor note, it's very important to make "moving some files around" and "changing those files" as separate commits because "diffing files against the previous state and noticing some are very similar" is exactly _how_ git notices some files moved. That's why in the process of rebasing this PR I introduced a pure "movement only" commit.

Yeah, sorry for that... move + small changes usually works fine but then the changes became bigger and bigger.

**Pull request merged**